### DISCUSSION TITLE
Compatibility with Microsoft Entra multifactor authentication external method provider

### DISCUSSION THREAD

**Discussion by nicolaspernoud** [ory/hydra#3909] [Q&A]
Hello, and thanks for this product !

Is it compatible with "Microsoft Entra multifactor authentication external method provider" (https://learn.microsoft.com/en-us/entra/identity/authentication/concept-authentication-external-method-provider) ?
The use case is to delegate the checking of another factor to the Idp, while using Entra Id as main IdP.

For this to work, there would be a few requirements :

- implement the implicit flow, or at least the id_token response type : https://learn.microsoft.com/en-us/entra/identity/authentication/concept-authentication-external-method-provider#microsoft-entra-id-call-to-the-external-identity-provider ;
- parse the id_token_hint provided by entra id : https://learn.microsoft.com/en-us/entra/identity/authentication/concept-authentication-external-method-provider#example-of-an-id_token_hint ;
- provide an id_token that must be an mfa token : https://learn.microsoft.com/en-us/entra/identity/authentication/concept-authentication-external-method-provider#example-of-an-eam-that-satisfies-mfa ;
- be extensible to integrate custom authentication methods.

**Comment by vinckr** [comment]
Hello @nicolaspernoud 
did you have a look at [Ory Kratos](https://github.com/ory/kratos)? As far as I can tell this is already possible with Ory Kratos (except custom auth methods, but I am not sure how that is related?)

**Reply by nicolaspernoud** [reply]
Hello @vinckr,
I had a look. It does not seems that kratos implements the required workflow that uses an id_token_hint emitted by an external provider (entra)... admittedly, Microsoft bend the OIDC protocol, that token should not be used that way according to the RFC.

**Reply by vinckr** [reply]
I see yea. 
Would love to see the ability to use external IdPs "MFA status" to skip additional MFA for the user - this is mainly what you are looking or right? 

So if the user has done MFA on their Entra you want to skip MFA in your system - did I get that correct? 
The complexity I can see is that every Idp interprets the RFC differently and thus it needs a custom implementation for every provider.

**Reply by nicolaspernoud** [reply]
Actually that is quite the opposite, I would like to delegate a second factor from entra to an IdP.

**Reply by vinckr** [reply]
Can you walk me through what you want to achieve? 

You want to skip MFA in entra if the user uses MFA in your system? 

edit: I think I got it from this description in the docs you shared
>An external authentication provider can integrate with Microsoft Entra ID tenants as an external authentication method (EAM). An EAM can satisfy the second factor of an MFA requirement for access to a resource or application.

**Reply by nicolaspernoud** [reply]
Exactly, the full workflow can be found here : https://learn.microsoft.com/en-us/entra/identity/authentication/concept-authentication-external-method-provider .
The point would be to have an external auth provider that should be customizable enough to have a custom factor, for example the ip address geolocation, of the presence of a VPN, etc.

**Reply by vinckr** [reply]
Hello @nicolaspernoud 

I see. 

>The point would be to have an external auth provider that should be customizable enough to have a custom factor, for example the ip address geolocation, of the presence of a VPN, etc.

You can already implement this behaviour in Ory Kratos via Ory Actions: https://www.ory.sh/docs/guides/integrate-with-ory-cloud-through-webhooks

Here is an example for VPN detection with Ory Actions: https://github.com/ory/examples/tree/master/ory-actions/vpncheck-py

Ory Network also gives you a `devices` object as part of the session that includes geolocation, IP address etc. so you can use those in Ory Actions. This is not directly implemented in Ory Kratos, but a proprietary feature.

Is this what you are looking for?

**Reply by nicolaspernoud** [reply]
That could be kind of what I am looking for : are Ory Actions part of the community version ?

**Reply by vinckr** [reply]
Yes you can use Ory Actions when self-hosting the OSS version of Ory Kratos.

**Reply by nicolaspernoud** [reply]
Thanks, I'll have a look !
